cs_volume.py

```diff
@@ -0,0 +1,749 @@
+            }
+            tags_to_create.append(tag)
+        if tags_to_create:
+            self.result['changed'] = True
+            if not self.module.check_mode:
+                args = {}
+                args['resourceids'] = resource['id']
+                args['resourcetype'] = resource_type
+                args['tags'] = tags_to_create
+                self.cs.createTags(**args)
+
+    def ensure_tags(self, resource, resource_type=None):
+        if not resource_type or not resource:
+            self.module.fail_json(msg="Error: Missing resource or resource_type for tags.")
+
+        if 'tags' in resource:
+            tags = self.module.params.get('tags')
+            if tags is not None:
+                self._delete_tags(resource, resource_type, tags)
+                self._create_tags(resource, resource_type, tags)
+                resource['tags'] = self.get_tags(resource)
+        return resource
+
+    def get_capabilities(self, key=None):
+        if self.capabilities:
+            return self._get_by_key(key, self.capabilities)
+        capabilities = self.cs.listCapabilities()
+        self.capabilities = capabilities['capability']
+        return self._get_by_key(key, self.capabilities)
+
+    def get_disk_offering(self, key=None):
+        disk_offering = self.module.params.get('disk_offering')
+
+        if not disk_offering:
+            return None
+
+        disk_offerings = self.cs.listDiskOfferings()
+        if disk_offerings:
+            for d in disk_offerings['diskoffering']:
+                if disk_offering in [d['displaytext'], d['name'], d['id']]:
+                    return self._get_by_key(key, d)
+        self.module.fail_json(msg="Disk offering '%s' not found" % disk_offering)
+
+    # TODO: for backward compatibility only, remove if not used anymore
+    def _poll_job(self, job=None, key=None):
+        return self.poll_job(job=job, key=key)
+
+    def poll_job(self, job=None, key=None):
+        if 'jobid' in job:
+            while True:
+                res = self.cs.queryAsyncJobResult(jobid=job['jobid'])
+                if res['jobstatus'] != 0 and 'jobresult' in res:
+                    if 'errortext' in res['jobresult']:
+                        self.module.fail_json(msg="Failed: '%s'" % res['jobresult']['errortext'])
+                    if key and key in res['jobresult']:
+                        job = res['jobresult'][key]
+                    break
+                time.sleep(2)
+        return job
+
+
+class AnsibleCloudStackVolume(AnsibleCloudStack):
+    def __init__(self, module):
+        AnsibleCloudStack.__init__(self, module)
+        self.volume = None
+
+    def get_volume(self):
+        if not self.volume:
+            args = {}
+            args['account'] = self.get_account(key='name')
+            args['domainid'] = self.get_domain(key='id')
+            args['projectid'] = self.get_project(key='id')
+            args['listall'] = True
```

Comment:
IMHO `listall=true` is not needed here. I try to avoid `listall=true` because if you run with root admin permissions, you will get more than you want and mix things, names are not always unique.

Maybe it would make sense to add `isrecursive=true` here instead.